engine/verification/fetcher/engine.go

```diff
@@ -190,6 +194,13 @@ func (e *Engine) processAssignedChunk(chunk *flow.Chunk, result *flow.ExecutionR
 		return false, blockHeight, nil
 	}
 
+	// skip chunk if it verifies a block at or above stop height
+	if e.stopAtHeight > 0 && blockHeight >= e.stopAtHeight {
+		e.log.Warn().Msgf("Skipping chunk %s - height  %d at or above stop height requested (%d)", chunkID, blockHeight, e.stopAtHeight)
+		e.chunkConsumerNotifier.Notify(chunkLocatorID) // tells consumer that we are done with this chunk.
```

Comment:
a bit concerned here, because if we notify consumer this chunk is done. Then after restart with a new version, this chunk will not be processed again. 

Technically, if all ENs stop at the same height, then no result above e.stopAtHeight will be received by VN. 

However, if some EN didn't set the stop height, and still send result to the VN, then VN will mark it has been processed, which create a state where the VN is assigned with a chunk, and has verified this chunk without sending approvals. And after VN is restarted, it will not process the chunks from the result that is received before the crash.

Reply:
Hmmm, I might miss something, but I think this state is not persisted, so restart will clean it

Reply:
This is actually persisted. 

`chunkConsumerNotify.Notify` will call `chunkconsumer.Worker.Notify`, which will call `consumer.NotifyJobIsDone`. 
And jobqueue's `NotifyJobIsDone`, might [call `SetProcessedIndex ` ](https://github.com/onflow/flow-go/blob/master/module/jobqueue/consumer.go#L239), which will [update the processed index in database](https://github.com/onflow/flow-go/blob/master/storage/badger/consumer_progress.go#L43), because VN uses a [consumer progress module to store processed index](https://github.com/onflow/flow-go/blob/master/cmd/verification_builder.go#L151) in database.

Reply:
Hmmm, does it simply looks like the solution is to omit this call to `e.chunkConsumerNotifier.Notify(chunkLocatorID)`?

Reply:
If we don't Notify the job is done, then this worker will never be recycled to work on a new job. VN has 5 workers in total, so if we got 5 chunks for block 100, and not notify they are done, then when chunk block 99 is received, we will have no worker to process it, and we will get stuck here. 

What we need is a way to only not notify the next chunk after block 99 is done, and notify other chunks after that chunk to be done .

This is a bit tricky, because the next chunk after block 99 is not always the first chunk in block 100, because block 100 might not have any chunk assigned to this VN.

Reply:
I think we can keep your change here as is. 

Instead, we could change the assigner engine to not process height above stopHeight. 

Assigner engine process each finalized block, and check whether there is chunk assigned to me. It [stores the finalized height as processed index](https://github.com/onflow/flow-go/blob/master/cmd/verification_builder.go#L293). In other words, we can learn what's the last processed height from the processed index. 

Since assigner process finalized block height by height in a single thread manner, if assigner engine is processing height 100, it means it must have finish processing all height below 100. In other words, if assigner is processing stopHeight, it means it have processed all height below that. If we don't notify 100 is done, it is ok, because all heights below 100 have been processed. 

But when to crash the VN? 
We want to stop VN when all chunks below height 100 that are assigned to this VN have been processed (verified). 
It is not straightforward to know because the VN might not get assigned with a chunk for every height, there might be no chunk assigned for block 99. 

If we know what's the chunk locator for the last chunk I got assigned before block 100, then we could just wait until the processed index reaches the last chunk to stop VN. 

One solution would be to let assigner engine tell fetcher engine what's the last chunk, and fetcher engine will stop when the processed index has reached the last chunk.